Rework NaN handling in sensor filters

DAVe3283 opened the pull request:
# What does this implement/fix?

tl;dr all sensor filters now handle NaN in a reasonable manner

While most filters would handle NaN (unknown) values correctly, the following would silently drop NaN values, preventing anything using the sensor from being informed of a problem: [`quantile`](https://esphome.io/components/sensor/index.html#quantile), [`median`](https://esphome.io/components/sensor/index.html#median), [`min`](https://esphome.io/components/sensor/index.html#min), [`max`](https://esphome.io/components/sensor/index.html#max), [`sliding_window_moving_average`](https://esphome.io/components/sensor/index.html#sliding-window-moving-average), [`exponential_moving_average`](https://esphome.io/components/sensor/index.html#exponential-moving-average), [`delta`](https://esphome.io/components/sensor/index.html#delta)

This could have physical consequences, e.g. imagine a temperature sensor controlling a heater in a fish tank where the sensor fails while the heater is on. The heater would never turn off, because the temperature output would never update or become unavailable.

~~I don't think this is a breaking change because the new behavior appears to be the expected behavior (see the related issues below). For those that want NaN to be dropped, adding a `filter_out: NaN` before other filter(s) will drop it while being explicitly clear about the intention. I suspect most people are already doing this if they want it.~~
The change in sensor behavior is significant enough this is a breaking change (see comments below). 

I didn't add myself as a codeowner but can if desired.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

### Breaking Change

[Sensor Filters](https://esphome.io/components/sensor/index.html#sensor-filters) now all handle NaN (unknown) values. If you used one of the following filters, this is a breaking change: [`quantile`](https://esphome.io/components/sensor/index.html#quantile), [`median`](https://esphome.io/components/sensor/index.html#median), [`min`](https://esphome.io/components/sensor/index.html#min), [`max`](https://esphome.io/components/sensor/index.html#max), [`sliding_window_moving_average`](https://esphome.io/components/sensor/index.html#sliding-window-moving-average), [`exponential_moving_average`](https://esphome.io/components/sensor/index.html#exponential-moving-average), [`delta`](https://esphome.io/components/sensor/index.html#delta)

To restore the previous behavior, add `filter_out: NaN` **before** the sensor filter. This will explicitly drop incoming NaN (unknown) values, making it clear what is happening and that it is desired.

**Related issue or feature (if applicable):** fixes esphome/issues#3140, fixes esphome/issues#3338, esphome/issues#3378, 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** N/A but I can make one if anyone feels the documentation needs clarification on NaN handling.

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [X] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yamlfilter_out: NAN
sensor:
  - platform: max31865
    name: "Air Temperature"
    cs_pin: GPIO15
    reference_resistance: 4.3 kΩ
    rtd_nominal_resistance: 1 kΩ
    rtd_wires: 3
    update_interval: 5s
    filters:
      # I verified the correct behavior of the fixed code on the following (one at a time)
      # - quantile
      # - median
      # - min
      # - max
      - sliding_window_moving_average:
          window_size: 5
          send_every: 3
      # - exponential_moving_average
      # - delta: 1.0
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

DAVe3283 committed:
Handle NaN in MedianFilter

DAVe3283 committed:
Handle NaN in QuantileFilter

DAVe3283 committed:
Handle NaN in MinFilter

DAVe3283 committed:
Handle NaN in MaxFilter

DAVe3283 committed:
Handle NaN in SlidingWindowMovingAverageFilter

DAVe3283 committed:
Handle NaN in DeltaFilter

DAVe3283 committed:
Handle NaN in ExponentialMovingAverageFilter

OttoWinter commented:
I agree the current "ignore NAN completely" is not good in the sensor filters.

I do think this is a breaking change, as it clearly changes behavior of sensors (even if what was before was not ideal). Additionally, the breaking changes are the perfect place to point users to the `- filter_out: NAN` line they can use to restore the old behavior.

DAVe3283 commented:
OK, I updated the description to indicate it is a breaking change. I don't think I can add the label unless a bot picks it up automatically.

All the filters that use a queue will slowly purge out the valid values as they fill with NaN, and once they are entirely full of NaN they will report NaN. So the data might get a bit jittery as it transitions from working to unavailable, but I think that is acceptable given the input is failing to begin with.

Do you want me to make a PR for the documentation to indicate how each sensor handles NaN? Or will the note in the breaking changes section suffice?

OttoWinter commented on esphome/components/sensor/filter.cpp:
```diff
@@ -37,27 +37,34 @@ MedianFilter::MedianFilter(size_t window_size, size_t send_every, size_t send_fi
     if (!this->queue_.empty()) {
-      std::deque<float> median_queue = this->queue_;
+      // Copy queue without NaN values
+      std::deque<float> median_queue;
```
```suggestion
      std::vector<float> median_queue;
```

Nit: vector takes a bit less memory, and readability of using the copy constructor is no longer there

OttoWinter commented on esphome/components/sensor/filter.cpp:
```diff
@@ -37,27 +37,34 @@ MedianFilter::MedianFilter(size_t window_size, size_t send_every, size_t send_fi
+        }
+      }
+
       sort(median_queue.begin(), median_queue.end());
```
Note to self: we can replace this with a [`std::nth_element`](https://en.cppreference.com/w/cpp/algorithm/nth_element) which has better average complexity (but not for this PR) - same for `quantile` filter

OttoWinter commented on esphome/components/sensor/filter.cpp:
```diff
@@ -74,26 +81,33 @@ void QuantileFilter::set_send_every(size_t send_every) { this->send_every_ = sen
     if (!this->queue_.empty()) {
-      std::deque<float> quantile_queue = this->queue_;
+      // Copy queue without NaN values
+      std::deque<float> quantile_queue;
```
```suggestion
      std::vector<float> quantile_queue;
```

OttoWinter commented on esphome/components/sensor/filter.cpp:
```diff
@@ -108,20 +122,26 @@ MinFilter::MinFilter(size_t window_size, size_t send_every, size_t send_first_at
-    float min = 0.0f;
-    if (!this->queue_.empty()) {
-      std::deque<float>::iterator it = std::min_element(queue_.begin(), queue_.end());
+    // Copy queue without NaN values
+    std::deque<float> min_queue;
+    for (auto v : this->queue_) {
+      if (!std::isnan(v)) {
+        min_queue.push_back(v);
+      }
+    }
+
+    float min = NAN;
+    if (!min_queue.empty()) {
+      std::deque<float>::iterator it = std::min_element(min_queue.begin(), min_queue.end());
       min = *it;
     }
```
For the other filters, copying the queue makes sense (because it's kind of necessary); But for `min`/`max` let's try to be more memory efficient (for long queue sizes).
```suggestion
    float min = 0.0f;
    for (auto v : this->queue_) {
      if (!std::isnan(v)) {
        if (std::isnan(min)) min = v;
        else min = std::min(min, v);
      }
    }
```

DAVe3283 replied:
Great suggestion, should be less RAM and way faster too. Will commit this manually so the initial value is `NAN` instead of `0.0f`;

OttoWinter commented on esphome/components/sensor/filter.cpp:
```diff
@@ -137,20 +157,26 @@ MaxFilter::MaxFilter(size_t window_size, size_t send_every, size_t send_first_at
 void MaxFilter::set_send_every(size_t send_every) { this->send_every_ = send_every; }
 void MaxFilter::set_window_size(size_t window_size) { this->window_size_ = window_size; }
 optional<float> MaxFilter::new_value(float value) {
-  if (!std::isnan(value)) {
-    while (this->queue_.size() >= this->window_size_) {
-      this->queue_.pop_front();
-    }
-    this->queue_.push_back(value);
-    ESP_LOGVV(TAG, "MaxFilter(%p)::new_value(%f)", this, value);
+  while (this->queue_.size() >= this->window_size_) {
+    this->queue_.pop_front();
   }
+  this->queue_.push_back(value);
+  ESP_LOGVV(TAG, "MaxFilter(%p)::new_value(%f)", this, value);
 
   if (++this->send_at_ >= this->send_every_) {
     this->send_at_ = 0;
 
-    float max = 0.0f;
-    if (!this->queue_.empty()) {
-      std::deque<float>::iterator it = std::max_element(queue_.begin(), queue_.end());
+    // Copy queue without NaN values
+    std::deque<float> max_queue;
+    for (auto v : this->queue_) {
+      if (!std::isnan(v)) {
+        max_queue.push_back(v);
+      }
+    }
+
+    float max = NAN;
+    if (!max_queue.empty()) {
+      std::deque<float>::iterator it = std::max_element(max_queue.begin(), max_queue.end());
       max = *it;
     }
```
```suggestion
    float max = 0.0f;
    for (auto v : this->queue_) {
      if (!std::isnan(v)) {
        if (std::isnan(max)) max = v;
        else max = std::max(max, v);
      }
    }
```

OttoWinter commented on esphome/components/sensor/filter.cpp:
```diff
@@ -167,30 +193,27 @@ SlidingWindowMovingAverageFilter::SlidingWindowMovingAverageFilter(size_t window
+        sum += v;
+        valid_count++;
+      }
+    }
```
Can we restore the old behavior of only recomputing the whole sum every 10000 iterations?

The reason is that if you have a longer window size, all these floating point operations will eat up a _lot_ of CPU time (on ESPs floating point is done in software);

DAVe3283 replied:
I did some testing on a D1 Mini (ESP8266):

<details>
<summary>config.yaml</summary>

```yaml
sensor:
  - platform: max31865
    name: "MAX31865 Debug"
    cs_pin: GPIO15
    reference_resistance: 4.3 kΩ
    rtd_nominal_resistance: 1 kΩ
    rtd_wires: 3
    update_interval: 0.5s
    filters:
      - sliding_window_moving_average:
          window_size: 7200 # 15 minute window @ 2/sec!
          send_every: 60
```
</details>

<details>
<summary>filter.cpp</summary>

```diff
 if (++this->send_at_ >= this->send_every_) {
   this->send_at_ = 0;
 
   float sum = 0;
   size_t valid_count = 0;
+  const uint32_t start = micros();
   for (auto v : this->queue_) {
     if (!std::isnan(v)) {
       sum += v;
       valid_count++;
     }
   }
+  const uint32_t calc = micros() - start;
+  const float us_per_item = static_cast<float>(calc) / static_cast<float>(valid_count);
+  ESP_LOGI(TAG, "SlidingWindowMovingAverageFilter(%p)::new_value(%f) calculated sum %zu/%zu items in %lu μs (%f μs/item)",
+    this, value, valid_count, this->window_size_, calc, us_per_item);
 
   float average = NAN;
   if (valid_count) {
     average = sum / valid_count;
   }
 
   ESP_LOGVV(TAG, "SlidingWindowMovingAverageFilter(%p)::new_value(%f) SENDING %f", this, value, average);
   return average;
 }
```
</details>

```
[15:20:51][I][sensor.filter:220]: SlidingWindowMovingAverageFilter(0x3fff31fc)::new_value(22.462816) calculated sum 781/7200 items in 946 μs (1.211268 μs/item)
[15:29:21][I][sensor.filter:220]: SlidingWindowMovingAverageFilter(0x3fff31fc)::new_value(22.395084) calculated sum 1801/7200 items in 2156 μs (1.197113 μs/item)
[15:39:21][I][sensor.filter:220]: SlidingWindowMovingAverageFilter(0x3fff31fc)::new_value(22.530348) calculated sum 3001/7200 items in 3574 μs (1.190936 μs/item)
[15:49:21][I][sensor.filter:220]: SlidingWindowMovingAverageFilter(0x3fff31fc)::new_value(22.428949) calculated sum 4201/7200 items in 4992 μs (1.188288 μs/item)
```

That looks pretty stable at ~1¼ μs per item, which is near a dead ringer for 100 clocks per pass through the loop. Which sounds bad, except even at the dismally slow 80 MHz of an ESP8266, the [recommended max of 20-30 ms per component](https://github.com/esphome/esphome/blob/fe535939a3b151d1280d2eddf09213f5eb956153/esphome/core/component.cpp#L200) works out to a window_size of 16000-24000.

The ESP8266 has 81920 bytes of RAM. Assuming all of it was avaliable to the std::deque<float>, and it has no overhead, we could only have a window_size of 81920/4 = 20480. So the realistic limit is likely less than half of that.

A window_size of 10000 would calculate in 8 milliseconds. For what it is worth, my ESP8266 crashed shortly after 4200 entries in the queue when heap ran out, while taking under 5 ms to calculate the average.

I can still put it back if you prefer, but I personally feel the increased accuracy on the default window size is well worth giving up 1.25 microseconds per entry every time the filter outputs a value.

OttoWinter replied:
Thanks for testing it so thoroughly; Indeed the additions appear to be quite quick (I just had in my mind a test I did some time ago for the ct_clamp, in which floating point math was very slow, but that also included multiplications/square roots, which are likely more expensive); anyway, right this doesn't appear to be a problem after all :+1:

OttoWinter commented on esphome/components/sensor/filter.cpp:
```diff
@@ -203,16 +226,18 @@ optional<float> SlidingWindowMovingAverageFilter::new_value(float value) {
-      this->accumulator_ = (this->alpha_ * value) + (1.0f - this->alpha_) * this->accumulator_;
-    }
+  if (std::isnan(value)) {
+    return value;
```
This short-circuits the `send_at >= send_every` code; I think it's probably better if we keep that in, otherwise the behavior for this filter is inconsistent with the other filters

DAVe3283 replied:
Yeah, good point. I'll figure out how to handle that efficiently.

OttoWinter commented:
> or rather close to the old behavior,

I suppose the old behavior can be achieved by having the `filter_out: NAN` **before** the sensor filter (so the filter only receives the not-NAN values)

DAVe3283 committed:
Apply suggestions from code review

Co-authored-by: Otto Winter <otto@otto-winter.com>

DAVe3283 committed:
Compute min and max more efficiently

DAVe3283 committed:
Make SENDING lines consistent across all filters

Most filters would report the final value being sent as the parameter???
This is now more clear what is happening, and consistent for all filters.

DAVe3283 committed:
Respect send_every for ExponentialMovingAverageFilter

This has the side-effect of only reporting a NaN if the input that triggers a send is NaN.
Otherwise they are ignored. It does prevent a dead sensor from being masked, though.

DAVe3283 commented:
@OttoWinter I believe I have addressed everything, so long as you agree with [the results of my testing](#discussion_r911472374).

OttoWinter reviewed (approved):
Thanks! :+1:

OttoWinter commented:
Looks good to me :)

Just want to ping @jesserockz for a second opinion/check I didn't miss anything (since this might affect some more complex configs which didn't expect the NAN)

DAVe3283 commented:
@jesserockz have you had a chance to take a look?

I have been running this on a couple devices with no ill effects, so we have almost a month of testing on my end so far.

jesserockz reviewed (approved):
LGTM.

Could you add a section to the PR description or just another comment I can copy to release notes around the breaking change? 
Cheers

DAVe3283 commented:
> Could you add a section to the PR description or just another comment I can copy to release notes around the breaking change? Cheers

Updated the description with the following:

# Breaking Change

[Sensor Filters](https://esphome.io/components/sensor/index.html#sensor-filters) now all handle NaN (unknown) values. If you used one of the following filters, this is a breaking change: [`quantile`](https://esphome.io/components/sensor/index.html#quantile), [`median`](https://esphome.io/components/sensor/index.html#median), [`min`](https://esphome.io/components/sensor/index.html#min), [`max`](https://esphome.io/components/sensor/index.html#max), [`sliding_window_moving_average`](https://esphome.io/components/sensor/index.html#sliding-window-moving-average), [`exponential_moving_average`](https://esphome.io/components/sensor/index.html#exponential-moving-average), [`delta`](https://esphome.io/components/sensor/index.html#delta)

To restore the previous behavior, add `filter_out: NaN` **before** the sensor filter. This will explicitly drop incoming NaN (unknown) values, making it clear what is happening and that it is desired.
